petastorm/reader_impl/shuffling_buffer.py

```diff
@@ -0,0 +1,168 @@
+    @abc.abstractmethod
+    def add_many(self, items):
+        """Adds multiple items to the buffer.
+
+        :param items: items to be added to the shuffling buffer.
+        :return: None
+        """
+        pass
+
+    @abc.abstractmethod
+    def retrieve(self):
+        """Selects an item from the buffer and returns the item to the caller. The item is removed from the buffer.
+
+        :return: The selected item.
+        """
+        pass
+
+    @abc.abstractmethod
+    def can_add(self):
+        """Checks the state of the buffer and returns whether a new item can be added to the buffer at the time.
+
+        :return: A boolean indicating whether an item can be added to the buffer at the time.
+        """
+        pass
+
+    @abc.abstractmethod
+    def can_retrieve(self):
+        """Checks the state of the buffer and returns whether an item can be removed from the buffer..
+
+        :return: A boolean indicating whether an item can be returned from the buffer at the time.
+        """
+        pass
+
+    @abc.abstractproperty
+    def size(self):
+        """Returns the number of elements currently present in the buffer.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+    @abc.abstractmethod
+    def finish(self):
+        """Call this method when no more `enqueu_many` calls will be made.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+class NoopShufflingBuffer(ShufflingBufferBase):
+    """A 'no-operation' (noop) implementation of a shuffling buffer. Useful in cases where no shuffling is desired, such
+    as test scenarios or iterating over a dataset in a predeterministic order.
+    """
+    def __init__(self):
+        self.store = []
+
+    def add_many(self, items):
+        self.store.extend(items)
+
+    def retrieve(self):
+        return self.store.pop(0)
+
+    def can_retrieve(self):
+        return len(self.store) > 0
+
+    def can_add(self):
+        return True
+
+    @property
+    def size(self):
+        return len(self.store)
+
+    def finish(self):
+        pass
+
+class RandomShufflingBuffer(ShufflingBufferBase):
+    """
+    A random shuffling buffer implementation. Items can be added to the buffer and retrieved in a random order.
+    """
+    def __init__(self, shuffling_buffer_capacity, min_after_retrieve, extra_capacity=1000):
+        """Initializes a new ShufflingBuffer instance.
+
+        Items may be retrieved from the buffer once `min_after_dequeue` items were added to the queue
+        (indicated by `can_retrieve`).
+
+        Items may be added to the buffer as long as the amount of items in the buffer does not exceed
+        `shuffling_queue_capacity`.
+
+        The amount of items in the buffer may grow above `shuffling_buffer_capacity` (due to a call to `add_many`
+        with a list of items), but must remain under `extra_queue_capacity`.
```

Comment:
maybe mention that `extra_queue_capacity` should aim to be equivalent to the maximum number of items you will try to add at once

Reply:
Would it be simpler to have chage `can_add()` to be `can_add(length)` which will return whether you can add a certain size instead of having this extra space? I can see a problem where if you have a small shuffling_buffer_capacity or try to add a lot of items at once but that shouldnt really ever be the case anyways?

Reply:
1. Added a clarification:
> Should be set to the upper bound of the number of items that can be added in a single call to `add_many` (can be a loose bound).

2. `can_add(length)` would be tricky as I need to know in advance if I can start pulling results from the loading queue and have a guarantee that I'll have somewhere to store them.